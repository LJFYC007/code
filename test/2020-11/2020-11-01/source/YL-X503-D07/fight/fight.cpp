#include<bits/stdc++.h>

using namespace std;

typedef unsigned long long u64;
#define int long long
#define REP(i, a, b) for ( int i = (a), _end_ = (b); i <= _end_; ++ i ) 
#define mem(a) memset ( (a), 0, sizeof ( a ) ) 
#define str(a) strlen ( a ) 
#define pii pair<int, int>
typedef long long LL;

const int p=1145141923;
const int Mod = 1145141923;

int a[10000005],b[10000005];
int n,m,q,w;
u64 seed;

u64 rnd()
{
	u64 x = seed;
	x ^= x << 13;
	x ^= x >> 7;
	x ^= x << 17;
	return seed = x;
}

signed main(){
	freopen("fight.in","r",stdin);
	freopen("fight.out","w",stdout);
	scanf("%lld%lld%lld%llu%lld",&n,&m,&q,&seed,&w);
	for(int i=0;i<=n;++i)a[i]=rnd()%w;
	for(int i=0;i<=m;++i)b[i]=rnd()%w;
    REP(i, 1, m) b[i] = (b[i] + b[i - 1]) % Mod;
	int l,r;
    REP(o, 1, q)
    {
		scanf("%lld%lld", &l, &r);
        int ans = 0; 
        REP(i, 0, n)
        {
            int L = max(0ll, l - i), R = min(m, r - i);
            if ( L <= R ) ans = (ans + a[i] * (b[R] - (L >= 1 ? b[L - 1] : 0))) % Mod;
        }
        printf("%lld\n", (ans + Mod) % Mod);
	}
	return 0;
}
