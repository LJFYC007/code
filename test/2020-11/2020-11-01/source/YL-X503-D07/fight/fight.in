100 95 100 7789078474991974168 95
106 134
85 164
91 136
98 109
169 184
80 99
85 89
40 74
140 148
4 13
107 173
0 79
99 151
127 193
68 134
101 113
61 111
2 135
17 169
42 55
115 167
9 156
147 172
97 162
42 84
8 151
24 170
83 133
69 170
18 176
132 186
98 162
157 163
13 79
111 131
6 153
70 152
5 158
136 156
33 166
43 179
170 172
74 89
12 50
88 139
111 143
0 49
44 167
15 71
92 153
14 55
52 61
190 192
37 53
26 35
35 133
61 81
110 142
11 93
19 144
91 122
153 193
98 117
51 60
97 189
2 178
11 56
42 163
1 33
10 175
40 65
52 121
97 148
63 135
20 176
44 75
68 162
39 190
20 72
0 145
181 184
3 106
6 111
78 164
73 103
146 192
101 158
19 165
41 138
80 151
19 41
88 112
68 115
71 190
51 165
63 107
28 167
177 194
41 160
70 193
