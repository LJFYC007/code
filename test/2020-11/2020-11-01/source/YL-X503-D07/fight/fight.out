4999364
12537451
8464856
2511749
724642
4057856
994436
4699031
1091587
203760
8416835
7470125
8511313
5425880
12379322
2599594
9654538
17902946
20850292
1541375
6562774
20218167
2087440
9918421
6275812
19736722
20570127
9646221
15692800
21164994
4546424
9705279
533374
7270790
3607361
19979790
14195124
20437876
2409467
19751066
19489512
149706
2940528
2882452
9442684
5120896
2937799
18795595
5807253
10264503
3405000
1303965
36922
1779609
732405
16210432
3509497
5182390
10067391
18582863
6404853
2111768
4020219
1257583
11126210
21641815
3628132
18757535
1356453
21382455
3176659
12638973
8626755
13296322
21071907
4459795
15382546
20201880
5771498
19135805
142138
12975452
13912961
13795219
6128720
2891908
8728111
20537691
16281770
12346271
1663365
5200095
9309085
16099970
17914623
8571317
20170425
493596
18649542
16306892
